tests/core/data_layer/test_data_store.py

```diff
@@ -400,91 +401,59 @@ async def test_inserting_duplicate_key_fails(
         )
 
 
-@pytest.mark.xfail()
 @pytest.mark.asyncio()
 async def test_autoinsert_balances_from_scratch(data_store: DataStore, tree_id: bytes32) -> None:
-    expected = Program.to(
-        (
-            (
-                (
-                    (b"\x00", b"\x10\x00"),
-                    (b"\x01", b"\x11\x01"),
-                ),
-                (
-                    (b"\x02", b"\x12\x02"),
-                    (b"\x03", b"\x13\x03"),
-                ),
-            ),
-            (
-                (
-                    (b"\x04", b"\x14\x04"),
-                    (b"\x05", b"\x15\x05"),
-                ),
-                (
-                    (b"\x06", b"\x16\x06"),
-                    (b"\x07", b"\x17\x07"),
-                ),
-            ),
-        ),
-    )
-
-    for n in [0, 4, 2, 6, 1, 3, 5, 7]:
-        await data_store.autoinsert(
-            key=bytes([n]),
-            value=bytes([0x10 + n, n]),
-            tree_id=tree_id,
-        )
+    random = Random()
+    random.seed(100, version=2)
+    hint_keys_values: Dict[bytes, bytes] = {}
+    hashes = []
 
-    result = await data_store.get_tree_as_program(tree_id=tree_id)
+    for i in range(2000):
+        key = (i + 100).to_bytes(4, byteorder="big")
+        value = (i + 200).to_bytes(4, byteorder="big")
+        node_hash = await data_store.autoinsert(key, value, tree_id, hint_keys_values)
+        hashes.append(node_hash)
 
-    assert result == expected
+    avg_height = 0
+    for node_hash in hashes:
+        ancestors = await data_store.get_ancestors_optimized(node_hash, tree_id)
+        assert len(ancestors) <= 14
+        avg_height += len(ancestors)
+    assert math.ceil(avg_height // 2000) == 11
```

Comment:
`//` already guarantees an integer.  With the `math.ceil()` i'm guess you meant `/` here?  Though that will probably end up `== 12` instead.

I think the above note is an error that should be fixed.  Below is a non-mandatory alternative to consider.  `:]`

I really like fewer assertions and to do them on data that will print interesting diagnostics on failure.  In this case this will notify you both of all the problematic node hashes and their heights.  I'm not certain that the second assertion is exactly right, but I think it matches the intent?  Or maybe it's off by one.

```python
heights = {node_hash: await data_store.get_ancestors_optimized(node_hash, tree_id) for node_hash in hashes}
too_tall = {hash: height for hash.height in heights.items() if height > 14}
assert too_tall == {}
assert 10 <= statistics.mean(heights.values()) <= 11
```